LiteCore/tests/QueryTest.cc

```diff
@@ -145,6 +145,366 @@ N_WAY_TEST_CASE_METHOD(QueryTest, "Create/Delete Array Index (Multi-level)", "[Q
+    results.resize(0);
+    while ( e->next() ) { results.push_back(e->columns()[0]->asString().asString()); }
+
+    vector<string> expected2 = expected;
+    for ( const auto& a : expected_inserted ) expected2.push_back(a);
+    CHECK(results.size() == 56);  // 16 + 40
+    CHECK(expected2 == results);
+
+    //
+    // Test the deletion triggers
+    //
+
+    // We now have 7 documents, from doc001 to doc007
+    // Let's delete doc002 and doc005, doc007, with one soft, one hard, and one soft delete
+    // soft delete is done by setting the delete flag and hard delete is to delete it from the kv-store
+    deleteDoc("doc002"_sl, false);
+    deleteDoc("doc005"_sl, true);
+    deleteDoc("doc007"_sl, false);
+
+    query = store->compileQuery(queryStr, QueryLanguage::kN1QL);
+    e     = query->createEnumerator();
+    results.resize(0);
+    while ( e->next() ) { results.push_back(e->columns()[0]->asString().asString()); }
+    vector<string> expected3;
+    for ( const auto& a : expected2 ) {
+        if ( a.find("doc002-") == string::npos && a.find("doc005-") == string::npos
+             && a.find("doc007-") == string::npos )
+            expected3.push_back(a);
+    }
+    CHECK(expected3.size() == 32);
+    CHECK(results == expected3);
+
+    // Undo the soft delete of the last doc, doc007
+    undeleteDoc("doc007"_sl);
+    query = store->compileQuery(queryStr, QueryLanguage::kN1QL);
+    e     = query->createEnumerator();
+    results.resize(0);
+    while ( e->next() ) { results.push_back(e->columns()[0]->asString().asString()); }
+    // putting back doc007-
+    for ( const auto& a : expected2 ) {
+        if ( a.find("doc007-") != string::npos ) expected3.push_back(a);
+    }
+    CHECK(expected3.size() == 40);
+    CHECK(results == expected3);
+
+    //
+    // Test the update triggers
+    //
+
+    // We now have 5 docs: doc001, doc003, doc004, doc006, doc007
+    // Update doc006
+
+    string updateDocID{"doc006"};
+    string updateJson = jsonBuilder(updateDocID);
+    // '{"a":[{"b":[{"c":["doc006-a0-b0-c0","doc006-a0-b0-c1"]},{"c":["doc006-a0-b1-c0",
+    //  "doc006-a0-b1-c1"]}]},{"b":[{"c":["doc006-a1-b0-c0","doc006-a1-b0-c1"]},
+    //  {"c":["doc006-a1-b1-c0","doc006-a1-b1-c1"]}]}]}'
+    size_t         pos = 0;
+    vector<string> arrayElements;
+    // Change doc006-a0-b0-c0 to doc006-A0-B0-C0, lower case a to c to upper case.
+    while ( pos < updateJson.length() ) {
+        pos = updateJson.find("\"doc006-", pos);
+        if ( pos != string::npos ) {
+            auto pos2 = updateJson.find("\"", pos + 1);
+            Assert(pos2 != string::npos);
+            ++pos2;
+            for ( size_t i = pos + 8; i < pos2; ++i ) {
+                if ( 'a' <= updateJson[i] && updateJson[i] <= 'z' ) { updateJson[i] = 'A' + updateJson[i] - 'a'; }
+            }
+            arrayElements.push_back(updateJson.substr(pos + 1, pos2 - pos - 2));
+            pos = pos2;
+        }
+    }
```

Comment:
Instead of this loop to update string, maybe modify the json generator to specify the format (e.g. lowercase or uppercase)?

Reply:
little awkward. I want to change the case for the string literals, like "a0-b0-c0", but not the property names. So I just did it hard way. It's one time need.